### PR TITLE
Homework #03: Data structures

### PR DESCRIPTION
**Pull request by alexey-kozlenkov**
**`traverse` method concern:** 
I firstly misread task description and didn't pay attention that `traverse` was requested to return an array of values, so I implemented it as I knew it. Even though later I discovered the miss, I still prefer to leave it `void`-ish + `callback` style since it gives wider possibilities for usage.
For instance, I was able to use BFS with specific callback to solve `getColumns` method 🙂

### CODE REVIEW

File: homeworks/03-data-structure/src/binary-search-tree/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { BinarySearchTree } from 'binary-search-tree'
+
+/**
```
Comment (dex157) [inline homeworks/03-data-structure/src/binary-search-tree/index.test.ts:3]: Illustration is always good idea :)

File: homeworks/03-data-structure/src/binary-search-tree/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  },
+}
+
+describe('binary search tree', () => {
+  describe('#has', () => {
+    it('returns true if element in tree', () => {
+      const tree = new BinarySearchTree(stubTreeRoot)
+
+      expect(tree.has(10)).toBe(true)
+      expect(tree.has(5)).toBe(true)
```
Comment (dex157) [inline homeworks/03-data-structure/src/binary-search-tree/index.test.ts:35]: For sets of similar tests is more preferable to use `it.each` https://jestjs.io/docs/api#testeachtablename-fn-timeout

**Review by dex157 — COMMENTED** [review]
Very good work, clean, clear and readable!

File: homeworks/03-data-structure/src/binary-search-tree/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    },
+  },
+}
+
+describe('binary search tree', () => {
+  describe('#has', () => {
+    it('returns true if element in tree', () => {
+      const tree = new BinarySearchTree(stubTreeRoot)
+
+      expect(tree.has(10)).toBe(true)
```
[the remark this answers is not in the document]
Comment (alexey-kozlenkov) [reply homeworks/03-data-structure/src/binary-search-tree/index.test.ts:34]: AFAI know there is no such built-in matcher [in jest](https://jestjs.io/docs/expect). There is `toBeTruthy` but it's not exactly the one we need

File: homeworks/03-data-structure/src/types/binary-search-tree.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { IBinaryTree, TreeNode } from './binary-tree'
+
+export interface IBinarySearchTree extends IBinaryTree<number> {
```
[the remark this answers is not in the document]
Comment (alexey-kozlenkov) [reply homeworks/03-data-structure/src/types/binary-search-tree.ts:3]: Duly noted.
Don't like using it either, however task specified there should be an interface and a class, so it was hard to come up with different names for them.

**Pull request merged**